MANIFEST.in

```diff
@@ -0,0 +1,19 @@
+include CMakeLists.txt CITATION LICENSE NOTICE .gitmodules mypy.ini requirements.txt version.txt
+recursive-include android *.*
+recursive-include aten *.*
+recursive-include binaries *.*
+recursive-include c10 *.*
+recursive-include caffe2 *.*
+recursive-include cmake *.*
+recursive-include torch *.*
+recursive-include tools *.*
+recursive-include test *.*
+recursive-include docs *.*
+recursive-include ios *.*
+recursive-include third_party *
+recursive-include test *.*
+recursive-include benchmarks *.*
+recursive-include scripts *.*
+recursive-include mypy_plugins *.*
+recursive-include modules *.*
+global-exclude *.o *.so *.dylib *.a .git
```

Comment:
You'll probably want to ensure to not do this in a dirty tree.

E.g. what SciPy does: https://github.com/scipy/scipy/blob/master/setup.py#L202

- in the `sdist` command in `setup.py`, check if all submodules are up to date.
- In the release script, ensure a `git clean -xdf` or running it on a clean git clone.

Reply:
Agree about the clean, but not necessarily about keeping submodules uptodate, as one should be at free to package their repo in whatever state of dirtiness into their `sdist` and `bdist` packages.

Reply:
You can always just commit whatever you want to package on a branch (even a temporary one), right? I'm not sure what the use case is for packaging things that aren't  committed anywhere.

Up to you of course, I just know SciPy implemented it after shipping an incomplete release and we've never had a complaint about the check.

Reply:
I've extracted existing submodules validation logic and share it across 'sdist' and 'build_deps' steps, will add some of your suggestions in the followup up diff.